### PR TITLE
Fix haddocks copy bug (#1105) (nearly)

### PR DESCRIPTION
**Pull request by sjakobi**
_Please don't merge this yet!_

This is simply to share some ideas on this problem before I go to sleep. :)

At this point, `stack` does generate haddocks for the dependencies and copies them too.
But now it tends to hit the following problem:

```
~/s/tasty (master) $ stack haddock
clock-0.5.1: configure
clock-0.5.1: build
clock-0.5.1: haddock
clock-0.5.1: install
tasty-0.11.0.1: configure
tasty-0.11.0.1: build
tasty-0.11.0.1: haddock
tasty-0.11.0.1: install
ProInvalid package identifier: "builtin_rts"
gress: 2⏎  
```

I'm not sure where in the code this error happens. Are there any tricks to get a line number or a something similar?

Please note that ghc-pkg seems to handle `builtin_rts` in a rather peculiar way too:

```
~/src $ ghc-pkg --version 
GHC package manager version 7.8.4
~/src $ stack exec ghc-pkg -- --version
Run from outside a project, using implicit global project config
Using resolver: lts-3.10 from implicit global project's config file: /home/simon/.stack/global/stack.yaml
GHC package manager version 7.10.2

~/src $ ghc-pkg field base depends
depends: ghc-prim-0.3.1.0-a24f9c14c632d75b683d0f93283aea37
         integer-gmp-0.5.1.0-26579559b3647acf4f01d5edd9491a46 builtin_rts
~/src $ stack exec ghc-pkg -- field base depends
Run from outside a project, using implicit global project config
Using resolver: lts-3.10 from implicit global project's config file: /home/simon/.stack/global/stack.yaml
depends:
    builtin_rts ghc-prim-0.4.0.0-af16264bc80979d06e37ac63e3ba9a21
    integer-gmp-1.0.0.0-8e0f14d0262184533b417ca1f8b44482

~/src $ ghc-pkg field builtin_rts id
ghc-pkg: cannot parse 'builtin_rts' as a package identifier
~/src $ stack exec ghc-pkg -- field builtin_rts id
Run from outside a project, using implicit global project config
Using resolver: lts-3.10 from implicit global project's config file: /home/simon/.stack/global/stack.yaml
ghc-pkg: cannot parse 'builtin_rts' as a package identifier
```

Note the strange alignment when asking for the dependencies of `base`.

(Referencing #1105)

### CODE REVIEW

File: src/Stack/GhcPkg.hs [hunk 1]
```diff
@@ -116,16 +115,16 @@ findGhcPkgField
     => EnvOverride
     -> WhichCompiler
     -> [Path Abs Dir] -- ^ package databases
-    -> String -- ^ package identifier, or GhcPkgId
+    -> PackageName
```
Comment (sjakobi) [inline src/Stack/GhcPkg.hs]: Bug 1, I believe. ghc-pkg doesn't accept full ghc-pkg ids:

```
~/src $ ghc-pkg field containers depends
depends: array-0.5.0.0-470385a50d2b78598af85cfe9d988e1b
         base-4.7.0.2-bfd89587617e381ae01b8dd7b6c7f1c1
         deepseq-1.3.0.2-63a1ab91b7017a28bb5d04cb1b5d2d02
         ghc-prim-0.3.1.0-a24f9c14c632d75b683d0f93283aea37
~/src $ ghc-pkg field base-4.7.0.2-bfd89587617e381ae01b8dd7b6c7f1c1 name
ghc-pkg: cannot find package base-4.7.0.2
```

File: src/Stack/GhcPkg.hs [hunk 2]
```diff
@@ -165,17 +164,12 @@ findGhcPkgHaddockHtml :: (MonadIO m, MonadLogger m, MonadBaseControl IO m, Monad
...
                       -> [Path Abs Dir] -- ^ package databases
-                      -> String -- ^ PackageIdentifier or GhcPkgId
+                      -> PackageIdentifier
                       -> m (Maybe (PackageIdentifier, Path Abs Dir))
-findGhcPkgHaddockHtml menv wc pkgDbs ghcPkgId = do
-    mpath <- findGhcPkgField menv wc pkgDbs ghcPkgId "haddock-html"
-    mid <- findGhcPkgField menv wc pkgDbs ghcPkgId "id"
-    mversion <- findGhcPkgField menv wc pkgDbs ghcPkgId "version"
-    let mpkgId = PackageIdentifier
-            <$> (mid >>= parsePackageName . T.encodeUtf8)
```
Comment (sjakobi) [inline src/Stack/GhcPkg.hs]: Bug 2:

```
λ> :set -XOverloadedStrings 
λ> parsePackageName "foo-1.0-132654"
*** Exception: Invalid package name: "foo-1.0-132654"
```

**Comment by borsboom** [comment]
I'm actually right in the middle of a major refactor of related code (basically, most of Stack.GhcPkg will be going away) as a first step toward #1105.  At this point I'm not sure whether the changes in this PR will still be relevant once that's done, so I hope your time hasn't been wasted.

**Comment by sjakobi** [comment]
No worries, it was a lot of fun! I'm looking forward to your refactor!

**Pull request closed without merging**